table/column.go

```diff
@@ -691,7 +691,9 @@ func FillVirtualColumnValue(virtualRetTypes []*types.FieldType, virtualColumnInd
 			}
 			// Because the expression might return different type from
 			// the generated column, we should wrap a CAST on the result.
+			sctx.GetSessionVars().StmtCtx.InFillVirtualColumnValue = true
```

Comment:
Setting it here may cause a data race. FillVirtualColumnValue() can be called concurrently.

Reply:
hi @wjhuang2016 , Can you give me some advice on how to fix it? The core reason is the "ShouldClipToZero"，but i have no idea how to solve it properly.

Reply:
I'm afraid there is not a clean way to do it. But we can add a particular check after CastValue() in FillVirtualColumnValue. We can check its type and value, then clip it to 0 if necessary.

Reply:
@wjhuang2016 i'm not very sure how to check it as you mentioned. After CastValue()，it should be a valiad destination type value already，how can judge it properly?

Reply:
You can know its type before CastValue(), and validate its value after CastValue(). If it doesn't obey the `ShouldClipToZero`, then ClipToZero.